ykrt/src/compile/jitc_yk/aot_ir.rs

```diff
@@ -385,122 +376,160 @@ impl AotIRDisplay for Operand {
 #[deku_derive(DekuRead)]
-#[derive(Debug)]
-pub(crate) struct Instruction {
-    type_idx: TypeIdx,
-    opcode: Opcode,
-    #[deku(temp)]
-    num_operands: u32,
-    #[deku(count = "num_operands")]
-    operands: Vec<Operand>,
-    /// A variable name, only computed if the instruction is ever printed.
-    #[deku(skip)]
-    name: RefCell<Option<String>>,
+#[derive(Debug, strum_macros::Display)]
+#[repr(u8)]
+#[deku(type = "u8")]
+pub(crate) enum Instruction {
+    #[deku(id = "0")]
+    Nop,
+    #[deku(id = "1")]
+    Load { ptr: Operand, type_idx: TypeIdx },
+    #[deku(id = "2")]
+    Store { val: Operand, ptr: Operand },
+    #[deku(id = "3")]
+    Alloca { type_idx: TypeIdx, count: usize },
+    #[deku(id = "4")]
+    Call {
+        callee: FuncIdx,
+        #[deku(temp)]
+        num_args: u32,
+        #[deku(count = "num_args")]
+        args: Vec<Operand>,
+    },
+    #[deku(id = "5")]
+    Br,
+    #[deku(id = "6")]
+    CondBr {
+        cond: Operand,
+        true_bb: BBlockIdx,
+        false_bb: BBlockIdx,
+    },
+    #[deku(id = "7")]
+    ICmp {
+        type_idx: TypeIdx,
+        lhs: Operand,
+        pred: Predicate,
+        rhs: Operand,
+    },
+    #[deku(id = "8")]
+    Ret {
+        #[deku(temp)]
+        has_val: u8,
+        #[deku(cond = "*has_val != 0", default = "None")]
+        val: Option<Operand>,
+    },
+    #[deku(id = "9")]
+    InsertValue { agg: Operand, elem: Operand },
+    #[deku(id = "10")]
+    PtrAdd {
+        // FIXME: the type will always be `ptr`, so this field could be elided if we provide a way
+        // for us to find the pointer type index quickly.
+        type_idx: TypeIdx,
+        ptr: Operand,
+        off: Operand,
+    },
+    #[deku(id = "11")]
+    BinaryOp {
+        lhs: Operand,
+        binop: BinOp,
+        rhs: Operand,
+    },
+    #[deku(id = "255")]
+    Unimplemented(#[deku(until = "|v: &u8| *v == 0", map = "map_to_string")] String),
 }
 
 impl Instruction {
-    /// Returns the operand at the specified index. Panics if the index is out of bounds.
-    pub(crate) fn operand(&self, idx: usize) -> &Operand {
-        &self.operands[idx]
-    }
-
-    /// Return a slice of the remaining operands, starting from the index `from` (inclusive).
-    pub(crate) fn remaining_operands(&self, from: usize) -> &[Operand] {
-        &self.operands[from..]
-    }
-
-    pub(crate) fn opcode(&self) -> Opcode {
-        self.opcode
-    }
-
-    /// For a call instruction, return the callee.
+    /// Find the name of a local variable.
     ///
-    /// # Panics
+    /// This is used when stringifying the instruction.
     ///
-    /// Panics if the instruction isn't a call instruction.
-    pub(crate) fn callee(&self) -> FuncIdx {
-        debug_assert!(matches!(self.opcode, Opcode::Call));
-        let op = self.operand(0);
-        match op {
-            Operand::Func(func_idx) => *func_idx,
-            _ => panic!(),
+    /// FIXME: This is very slow and could be optimised.
```

Comment:
The value being computed here used to be computed and cached (post-serialise) inside every single instruction the defined a value. Since the API user would have to see that fluff, I've opted to remove the cache and compute it every time.

Since it's only used for stringification, I hope that's OK.

Reply:
(If we wanted to bring back the cache, we could use a hashmap in the module, but it'd require peppering lifetimes all over `Module` -- I didn't want to junk up this PR with that, and it may never really be an issue anyway)

Reply:
The performance of this isn't a problem.